Fix uncompress tgz no extension

danlipsa opened the pull request.

Commit:
Function to uncompress tgz does not need to check for tar extension

This check interferes with uncompressing a tgz without extension

Commit:
Add check to remove extension from a path without extension.

Without this check, an assumption that there is an extension
make the code abort

danlipsa commented:
@johnwparent Please review

johnwparent reviewed (dismissed):
Looks great, lgtm

johnwparent commented:
@scheibelp For some reason the fetcher strips the extension off of some archive files, and my recent changes to the compression module do not play well with this. This PR from @danlipsa fixes the compression module side of things.
I'm unsure if this is a larger bug in the fetcher as the behavior dates back to the 0.17.0 release (or at least that's as far back as I tested), so there may be some follow up effort needed to that end. Let me know if that's expected/ok or if I should dig in there.

scheibelp reviewed (changes requested):
The changes look great but can we add a unit test to avoid this sort of regression in the future? e.g. we could copy `lib/spack/spack/test/data/compression/Foo.gz` to an extension-less path and then create a test that reproduces the error that led to this PR (I'm assuming it would use resources in `test/url_fetch` and/or `test/util/compression`.

Commit:
Tar on windows cannot untar onto itself.

Commit:
Uncompressed txz needs to go to tar

danlipsa commented:
@scheibelp @johnwparent I added tests for extension-less compressed files as suggested, and fixed two more issues revealed by the test.  Please review.

Commit:
Add tests for extension less compressed files

Commit:
Fix style

danlipsa commented:
@johnwparent @scheibelp Here is a simpler fix for the same issue: just save the URL that originally did not have any extension (pegtl) to a file with the proper extension. This avoids all the issues we are hitting when uncompressing a file without an extension.
https://github.com/spack/spack/pull/38476

johnwparent commented:
@danlipsa, @scheibelp can feel free to correct me here if he disagrees, but regardless of the other PR you linked, I think we'd still want to support the ability to decompress and extract extension-less archives rather than just relying on a file rename. I think the changes in this PR fix some fragile behavior and should still make it into Spack. I'll wait on @scheibelp to reopen however.

danlipsa commented:
@johnwparent I am split on this. The only reason we got in trouble here is that pegtl provides its release source, which is tar.gz as an extension-less file - which is strange. Because we have code to save the file using the same name used in the release URL we end up with this issue. Not that for the cache we introspect the file and figure out the proper extension which we use for the file name there. So my new fix uses the same file name with extension as we use for the cache. While we could make things word for files without extensions, I don't think it is needed.

danlipsa commented:
But I defer to you guys if we need this functionality in spack.

danlipsa committed:
Un Unix gunizp cannot unzip an extensionless file

danlipsa commented:
@scheibelp For this PR all tests pass. Is this ready? Thanks!

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -77,8 +77,12 @@ def _system_untar(archive_file):
+    if archive_file_no_ext == archive_file:
+        # the archive file has no extension. Tar on windows cannot untar onto itself
+        archive_file = archive_file_no_ext + ".tar"
+        shutil.copy(archive_file_no_ext, archive_file)
```
I think this should be a move vs. a copy (I assume there's no issue in the sense that the untar overwrites the copy with the original name, but my issue is more with creating duplicates of the file when it isn't necessary). 

Likewise for the other functions modified in this PR in this way.

Also, I think it's worth choosing a modified name other than adding `.tar`, since the main significant point of this is to just choose a name different than the intended final unpacked directory. (something like `archive_file = archive_file_no_ext + "-input"`).

Likewise for the other functions modified in this PR in this way.

danlipsa replied:
> I think this should be a move vs. a copy (I assume there's no issue in the sense that the untar overwrites the copy with the original name, but my issue is more with creating duplicates of the file when it isn't necessary).

Sounds good.

danlipsa replied:
> Also, I think it's worth choosing a modified name other than adding .tar, since the main significant point of this is to just choose a name different than the intended final unpacked directory. (something like archive_file = archive_file_no_ext + "-input").

There are two problems I've seen: 1. tar on windows complains it cannot untar a file onto itself. 2. on unix unzip complains when there is no extension. So '-input' won't work on issue 2. Adding the proper extension solves the issues consistently.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -159,7 +163,12 @@ def _py_gunzip(archive_file):
-    decompressed_file = os.path.basename(strip_compression_extension(archive_file, "gz"))
+    archive_file_no_ext = strip_compression_extension(archive_file, "gz")
+    if archive_file_no_ext == archive_file:
+        # the zip file has no extension. Un Unix gunzip cannot unzip onto itself
```
Typo: "Un Unix"

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -233,22 +242,24 @@ def unarchive(archive_file):
-    decompressed_file = os.path.basename(strip_extension(archive_file, "xz"))
+    if check_extension(archive_file, "txz"):
+        decompressed_file = os.path.basename(strip_extension(archive_file, "txz"))
+        decompressed_file += ".tar"
```
Are there not similar issues for `.tgz`?

I assume an analogous change isn't necessary because of the changes in `_win_compressed_tarball_handler` (which no longer bails if the extension is not `.tar`); that also means that adding `decompressed_file += ".tar"` is unnecessary, although to be clear, I think stripping the `txz` extension is good.

I think other functions should copy the behavior of properly stripping the combined extensions (e.g. "tgz", "tbz"), since otherwise the intermediate decompressed (but not un-tar-ed) file retains the extension type, which could cause confusion when folks debug.

danlipsa replied:
> Are there not similar issues for .tgz?
On windows, this extension goes through _system_untar so it behaves different there.

> I assume an analogous change isn't necessary because of the changes in _win_compressed_tarball_handler (which no longer bails if the extension is not .tar); that also means that adding decompressed_file += ".tar" is unnecessary, although to be clear, I think stripping the txz extension is good.

Indeed, tests pass without .tar extension.

Commit:
Use move instead of copy and fix typo

Commit:
No need for the .tar extension to untar a file

danlipsa committed:
Avoid taking different extension as no extension

danlipsa commented:
@scheibelp Please review. All tests pass.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -77,8 +77,12 @@ def _system_untar(archive_file):
     """
-    outfile = os.path.basename(strip_extension(archive_file, "tar"))
-
+    archive_file_no_ext = strip_extension(archive_file)
```
I think this should still be `strip_extension(archive_file, "tar")`: while this won't strip arbitrary extensions like ".foo", ~it might still do the wrong thing if handed a file in an incorrect state (e.g. a `tar.gz`), and it would be confusing if we stripped off the `.gz`, then complained about not being able to untar something with a name ending in ".tar".~ (note, this was wrong and didn't account for the fact that the system untar generally handles .tgz, .tar.gz, and .tar, so stripping all extensions as per the method defined in this module is appropriate)

Likewise for the other calls to `strip_extension`.

danlipsa replied:
This was done initially, but I run into troubles with '*.tgz', which was not removed, so the following if was evaluated true. The file was judged a file with no extension, so I ended up with an archive file <name>.tgz.tar which was incorrect.
I think your example (*.tar.gz) would work fine. gz would be stripped, so the if statement would be false. The file won't be judged as having no extension.

scheibelp replied:
OK, and moreover I think my initial comment missed that this method can get more than just a ".tar" or ".tar.gz": at least the Linux `tar` utility can handle a `.tar.gz` that omits "tar" from the name (e.g. if you rename `foo.tar.gz` to `foo.gz`, then `tar` on Linux appears to be fine with that).

This does get into a discrepancy between Windows and Unix that leads to a conceptual conflict in this method:

* On Unix we could get all sorts of file types, so appending ".tar" is potentially misleading (e.g. if the file was a .tar.gz)
* On Windows, we'd only ever be getting a tar file (not a .tar.gz - some prior step would have handled that)

Therefore, I think only adding `.tar` on Windows would be ideal

danlipsa replied:
or adding '-input' would work as well. I think you suggested this at one point.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -159,7 +163,12 @@ def _py_gunzip(archive_file):
+    if archive_file_no_ext == archive_file:
+        # the zip file has no extension. On Unix gunzip cannot unzip onto itself
+        archive_file = archive_file + ".gz"
+        shutil.move(archive_file_no_ext, archive_file)
```
I think this PR should apply the same level of care to `bz2`, `xz`, etc.: if the filename is missing the extension, we add it.

danlipsa replied:
The reason why we need these renames is to go around issues or bugs with the utilities we use to decompress. In one case, tar on windows complains that it cannot uncompress onto itself. In a second case, unzip on unix complains about a file with no extension. While we test all possible extensions, we don't hit other issues because the utilities we use don't have the same drawbacks. So, for simplicity I propose to not address the other extensions as there is no fundamental reason why these renames are needed. It is only to go around bugs in the utilities.

scheibelp replied:
Fair enough. I was inclined to say that these specific cases inform us of good general rules that - while not strictly required in all cases - might help us avoid similar complaints in the future (i.e. there's no harm in appending `.xz` and we may avoid some Windows/Unix-implementation-specific complaint in the future) - not something that has to be addressed here though (and I think there is definitely some value in knowing the actual tools with the problem).

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -233,22 +242,23 @@ def unarchive(archive_file):
+    if check_extension(archive_file, "txz"):
+        decompressed_file = os.path.basename(strip_extension(archive_file, "txz"))
+    else:
+        decompressed_file = os.path.basename(strip_extension(archive_file, "xz"))
```
I think this PR should apply the same level of care to `.bz2`/`.tbz` etc. I recommend making a function to streamline that:

```
def prep_archive(archive_file, extensions):
  for ext in extensions:
      outfile = strip_extension(..., ext)
      if outfile != archive_file:
        break
  else:
    outfile = archive_file
    inferred_ext = next(iter(set.union(extensions, fs.EXTS)))
    renamed_archive_file += inferred_ext
    shutil.move(archive_file, renamed_archive_file)

...

prep_archive(archive_file, ["txz", "xz"])
```

Note this also incorporates the other generalization I requested in `_system_gunzip`

danlipsa replied:
Please see my previous comment. Let me know if you prefer to do these renames consistently. We can certainly do that.

scheibelp replied:
I agree with your comment on https://github.com/spack/spack/pull/37749/files#r1258633952, but I think this particular thread raises a concern which can be considered severable, and IMO that concern persists: for example I think `_py_gunzip` should not decompress `foo.tgz` to a file that is also named `foo.tgz` (regardless of whether it can in fact do that).

danlipsa replied:
I think that's a great observation that the affected functions should have reasonable outputs even if their input has no extension. That includes tgz should not decompress to tgz. I'll look at the input and output for the affected functions on Windows and Unix.

danlipsa replied:
your concern about gunzip applied to tgz is already handled by using strip_compression_archive which expands tgz to tar.gz and then gz is stripped, so the output file is correctly infered as tar.

danlipsa replied:
I changed txz to be handled the same way in py_lzma

Commit:
The input to _system_untar can be tar, tgz, tbz2 and others

Commit:
Use strip_compression_extension for txz

scheibelp commented on lib/spack/spack/test/util/compression.py:
```diff
@@ -39,6 +39,32 @@ def compr_support_check(monkeypatch):
 
 
+@pytest.fixture
+def archive_file_and_extension(tmpdir_factory, request):
```
Can you rewrite this to use the `archive_file` fixture? That should remove some duplication here (you still have to strip the extension off here).

danlipsa replied:
Indeed!

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -708,14 +715,16 @@ def extension_from_path(path):
+        for ext_check in exts_to_check:
+            mod_path = check_and_remove_ext(path, ext_check)
+            if mod_path != path:
+                return mod_path
```
This looks good but can you add a unit test which makes sure that e.g. `Foo.tgz` -> `Foo.tar`, etc.?

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -77,8 +77,12 @@ def _system_untar(archive_file):
+    outfile = os.path.basename(archive_file_no_ext)
+    if archive_file_no_ext == archive_file:
+        # the archive file has no extension. Tar on windows cannot untar onto itself
+        archive_file = archive_file_no_ext + "-input"
```
IMO it's worth commenting that on Unix, this function may receive files that are not just `.tar` (e.g. `.tar.gz`), so naming this `-input` avoids implying this is any specific type of archive.

scheibelp reviewed (changes requested):
This looks good but I have a few requests about testing and comments.

Commit:
Expand comment to explain why we add -input postfix

Commit:
Refactor using parameterize to avoid code duplication

Commit:
Add test for strip_compression_extension

Commit:
Add additional test for specified zip extension

Commit:
Fix style

danlipsa committed:
Fix archive_file_and_extension on Unix

danlipsa committed:
Fix style

danlipsa commented:
@scheibelp Please review, I addressed all your comments. Thanks!

scheibelp commented:
Thanks!

haampie commented:
0.20.1 material?

scheibelp commented:
Yes
